### PR TITLE
PHPStan 2.0 support

### PR DESCRIPTION
**Pull request by swissspidy**
🤞 

- Bumps PHPStan requirement to 2.0+
- Bumps PHP requirement to 7.4+ accordingly
- Follows upgrade guide to pass new arguments where needed

There are currently 13 failing tests, all regarding `DynamicReturnTypeExtension`.

See #242

### CODE REVIEW

**Comment by swissspidy** [comment]
> There are currently 13 failing tests, all regarding `DynamicReturnTypeExtension`.

Any volunteers?

**Comment by swissspidy** [comment]
Many failures are related to `EchoKeyDynamicFunctionReturnTypeExtension` specifically. Seems like the `::getDefaultReturnType` method doesn't return the correct default return type here:

https://github.com/szepeviktor/phpstan-wordpress/blob/fdc31b2a37515acbe03069baf46489f88d0a4821/src/EchoKeyDynamicFunctionReturnTypeExtension.php#L181-L185

For example, for `wp_get_archives` it should be `string|void` as it can output both, but I only get `void`.

In the stubs the definition for that function is `@phpstan-return ($args is array{echo: false|0} ? string : void)`.

Maybe that's incorrect and it should be `@phpstan-return ($args is array{echo: false|0} ? string|void : void)` though? It's one of the functions that can always potentially return `void`, even if `echo=false`.

**Comment by IanDelMar** [comment]
> Any volunteers?

I'll have a look at this.

**Comment by szepeviktor** [comment]
So this should be switched to `2.x` branch.

**Comment by IanDelMar** [comment]
The test for `_get_list_table()` fails due to the current version of the stubs. As of version 6.6.0, the stubs do not permit unknown types to be passed to `_get_list_table()`. An unknown type, or more specifically, any type other than `class-string`, will result in a PHPStan error concerning the passed argument.

**Comment by swissspidy** [comment]
> So this should be switched to `2.x` branch.

Done (I hope)

@IanDelMar feel free to make/suggest any code changes

File: src/HookDocsRule.php [hunk 1]
```diff
@@ -216,7 +216,7 @@ protected function validateSingleParamTag(string $paramName, ParamTag $paramTag,
             $this->currentScope->isDeclareStrictTypes()
         );
 
-        if (! $accepted) {
+        if ($accepted->result) {
```
Comment (swissspidy) [inline src/HookDocsRule.php:219]:
```suggestion
        if (! $accepted->result) {
```

I think. Oversight from rebasing?

File: src/AssertWpErrorTypeSpecifyingExtension.php [hunk 2]
```diff
@@ -38,7 +38,12 @@ public function specifyTypes(MethodReflection $methodReflection, MethodCall $nod
     {
         $args = $node->getArgs();
 
-        return $this->typeSpecifier->create($args[0]->value, new ObjectType('WP_Error'), TypeSpecifierContext::createTruthy());
+        return $this->typeSpecifier->create(
+            $args[0]->value,
+            new ObjectType(\WP_Error::class),
```
Comment (swissspidy) [inline src/AssertWpErrorTypeSpecifyingExtension.php:43]: Or is `ObjectType('WP_Error')` preferred?

**Comment by szepeviktor** [comment]
This is so COMPLEX!
I hope everything has a **test**.

File: src/AssertWpErrorTypeSpecifyingExtension.php (hunk 2 above)
Comment (szepeviktor) [reply src/AssertWpErrorTypeSpecifyingExtension.php:43]: Less strings -> brighter future 🌻

**Comment by IanDelMar** [comment]
@swissspidy You need to remove PHP 7.2 from travis.yml

**Comment by szepeviktor** [comment]
We have only a few problems in CI.

**Review by szepeviktor — APPROVED** [review]

**Comment by IanDelMar** [comment]
PR with fix for `wp_get_archives()` in https://github.com/php-stubs/wordpress-stubs/

**Comment by IanDelMar** [comment]
> We have only a few problems in CI. Very few.

I guess that most or all of them would be resolved if we merge #235

**Comment by szepeviktor** [comment]
> I guess that most or all of them would be resolved if we merge #235

Done 🍏

**Comment by IanDelMar** [comment]
@swissspidy, could you rebase your PR so we can see what still needs to be resolved?

**Comment by IanDelMar** [comment]
All errors, except for a few cases related to `wp_get_archives()` that have already been resolved in the stubs, arise from restricting argument types in the stubs in combination with conditional return types that modify the default return value. If the argument types are restricted in the stubs, then—knowing that PHPStan will handle this—the stubs do not need to account for the return type of an invalid argument.

For `_get_list_tables()`, see [this comment](https://github.com/szepeviktor/phpstan-wordpress/pull/243#issuecomment-2403482830). The other errors result from restricting the argument to arrays.

Testing with an invalid argument type, such as a query string, may lead to unexpected return types depending on the default return value of the conditional return type (i.e. the `else` branch that is reached if none of the `if` or `elseif` conditions are met). If PHPStan encounters `'echo=1&akey=avalue'`, it is considered a string, and the conditional return types `($args is array{echo: false|0} ? string : void)` and `($args is array{echo: true|1} ? string : void)` both evaluate to false (string ≠ array). The solution is to avoid testing with invalid argument types.

This issue does not occur with all functions with restricted arguments in the `echo` key extension because not all extensions have been ported to the stubs (i.e. not all have conditional return types). Some were incorrect (I submitted PRs to remove these extensions), and some have not been ported because `$args` is filterable, meaning we cannot determine the actual value or type of `$args`.

As there is no `echo` key extension left in v2.x, we only need to address this in the master branch. I suggest removing the tests that fail due to what I have just described.

**Comment by szepeviktor** [comment]
We have done a lot work.
Ian: keyboard
Viktor: mouse

File: tests/HookDocsRuleTest.php [hunk 3]
```diff
@@ -26,6 +26,7 @@ protected function getRule(): Rule
         return new HookDocsRule($fileTypeMapper, $ruleLevelHelper);
     }
 
+    // phpcs:ignore SlevomatCodingStandard.Functions.FunctionLength.FunctionLength
```
Comment (IanDelMar) [inline tests/HookDocsRuleTest.php:29]:
```suggestion
-    // phpcs:ignore SlevomatCodingStandard.Functions.FunctionLength.FunctionLength
```

File: tests/HookCallbackRuleTest.php [hunk 4]
```diff
@@ -17,6 +17,7 @@ protected function getRule(): \PHPStan\Rules\Rule
         return new HookCallbackRule();
     }
 
+    // phpcs:ignore SlevomatCodingStandard.Functions.FunctionLength.FunctionLength
```
Comment (IanDelMar) [inline tests/HookCallbackRuleTest.php]: This has been moved to the config file and can be removed here.

File: tests/HookDocsRuleTest.php (hunk 3 above)
Comment (IanDelMar) [inline tests/HookDocsRuleTest.php]: This has been moved to the config file and can be removed here.

**Comment by szepeviktor** [comment]
Are we ready in this PR?

**Comment by swissspidy** [comment]
I think so. No idea what's wrong with that `Codacy Static Code Analysis` check though.

**Comment by IanDelMar** [comment]
`stdout: ERROR: Referenced sniff "PSR12NeutronRuleset" does not exist` :open_mouth:

**Pull request merged**